### PR TITLE
@ #121 | How to use Docker in Docker (DinD) and Docker outside of Docker (DooD) for local CI testing

### LINKED ISSUE #121: should create a blog post: "How to use Docker in Docker (DinD) and Docker outside of Docker (DooD) for local CI testing"
- the differences
- the use cases
- tips
  + volume mount
  + ...?

### PR DESCRIPTION
**Pull request by hoatle**
connect #121 

@hoavt we need a technical reviewer for this, too.

### CODE REVIEW

**Comment by hoavt** [comment]
@hieptranquoc: Please help to review the technical aspect of this PR today.

**Comment by hoavt** [comment]
Assigned to @datphan to review :)

File: source/_posts/2017-09-11-how-to-use-docker-in-docker-dind-and-docker-outside-of-docker-dood-for-local-ci-testing.markdown [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+DinD includes a whole Docker installation inside of it.
+
+DooD uses its underlying host's Docker installation by bind-mounting the Docker socket.
+
+Obviously, DooD should be faster because we can leverage its caching mechanism, and DinD should be
+cleaner. DinD should support parallel running but DooD does not, or at least, not very reliable
+with my observation. If you want to conduct the clean testing, use DinD. If you want to conduct the
+faster testing, use DooD.
+
+Dood is simpler to use than DinD.
```
Comment (datphan) [inline source/_posts/2017-09-11-how-to-use-docker-in-docker-dind-and-docker-outside-of-docker-dood-for-local-ci-testing.markdown]: DooD

File: source/_posts/2017-09-11-how-to-use-docker-in-docker-dind-and-docker-outside-of-docker-dood-for-local-ci-testing.markdown [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+Although running Docker inside Docker (DinD) or Docker outside of Docker (DooD) is generally not
+recommended, there are some legitimate use cases, such as development of Docker itself or for local
+CI testing. And in this blog post, I'm going to show you how to use DinD or DooD for local CI testing
+as it's a very typical use case for local DevOps.
+
+<!-- more -->
+
+## The differences between DinD and DooD?
+
+DooD is the opposite of DinD.
```
Comment (datphan) [inline source/_posts/2017-09-11-how-to-use-docker-in-docker-dind-and-docker-outside-of-docker-dood-for-local-ci-testing.markdown]: DinD is the opposite of DooD.

**Comment by hoatle** [comment]
@datphan @hoavt feedbacks applied

**Comment by datphan** [comment]
cofirmed

File: source/_posts/2017-09-11-how-to-use-docker-in-docker-dind-and-docker-outside-of-docker-dood-for-local-ci-testing.markdown [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+CI testing. And in this blog post, I'm going to show you how to use DinD or DooD for local CI testing
+as it's a very typical use case for local DevOps.
+
+<!-- more -->
+
+## The differences between DinD and DooD?
+
+DinD is the opposite of DooD.
+
+DinD includes a whole Docker installation inside of it.
```
Comment (hoavt) [inline source/_posts/2017-09-11-how-to-use-docker-in-docker-dind-and-docker-outside-of-docker-dood-for-local-ci-testing.markdown]: --> DinD includes the whole Docker installation inside it.

File: source/_posts/2017-09-11-how-to-use-docker-in-docker-dind-and-docker-outside-of-docker-dood-for-local-ci-testing.markdown [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+```
+
+Let's see how to do local CI testing in action:
+
+<script type="text/javascript" src="https://asciinema.org/a/137314.js" id="asciicast-137314" async></script>
+
+
+This is just the very first step for basic testing.
+
+Later, we should convert this `.travis.yml` file into a `build.sh` to execute, this is the right way
```
Comment (hoavt) [inline source/_posts/2017-09-11-how-to-use-docker-in-docker-dind-and-docker-outside-of-docker-dood-for-local-ci-testing.markdown]: a `build.sh` to --> the `build.sh` one to execute. This is the...

File: source/_posts/2017-09-11-how-to-use-docker-in-docker-dind-and-docker-outside-of-docker-dood-for-local-ci-testing.markdown (hunk 3 above)
Comment (hoatle) [reply source/_posts/2017-09-11-how-to-use-docker-in-docker-dind-and-docker-outside-of-docker-dood-for-local-ci-testing.markdown]: I keep this

**Comment by hoatle** [comment]
@hoavt feedbacks applied, thank you.

**Pull request merged**